### ISSUE TITLE
flags.midi-qol.optional broken ?

### ISSUE THREAD

**Issue by tposney** [tposney/midi-qol#1103]
In GitLab by @google153 on Dec 24, 2022, 20:34

** updated see latest comment **

Is the midi-qol.optional flag currently broken ?

Yes, i know stupid users we are, and no you do not need to fix my computer :-). 
But i already tried alot and got stucked so i wanted to check if maybe the feature is broken, before i waste another day.

I tried own effects using:
![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/161edf86d1cbaaf04157255df17d5625a8e54a840175e49c43ddfd7d9c64bab9/image.png)

I am using default settings full automation:
[fvtt-midi-qol-settings.json](https://my-gitlab-bucket.s3.amazonaws.com/640711674/f46fdffc7ae91cd5427aea3a60e9cf510d5a73b4e90426e8dcd63d690b82a06c/fvtt-midi-qol-settings.json)

I reduced the modules running to minimal: all up to date at the time written.
![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/2068d032ce93bd69a4c7729f0ef203bb6b7cf30ebcc29a36de9ffa1bed9cac88/image.png)
![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/e24fa3813a81386763fb27c337edb64382491c5efdb3ca60043b7eec5d154653/image.png)

**Comment by tposney** [comment]
In GitLab by @google153 on Dec 24, 2022, 23:38

Well it is working. 
After looking into the source code i found my problem with lucky it needed active usages on actor left.
However i currently did not find any sourcecode occurence for optional in combination with ac. But i probably will find out.

**Comment by tposney** [comment]
In GitLab by @google153 on Dec 25, 2022, 24:33

First of one has to understand that optional.ac is handled as reaction.
but then after some debugging i found one bug: 

if the user do not own any real reaction it will not process.

**in utility.js : doReactions, line 2804:** 

![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/15c2aa996ee9ef3773889cfa4d2697f909b00ee11f748f67d61e997045534e99/image.png)

it seems you did a refactoring typo here at some point. originally you wanted obviously store a reactions object and not a reactionCount object. Now you store within reactions the count including the optionals.

later on you check for reactionCount wich still is zero. while reactions is now 1.



Also i did not yet figgure out why it then still does not work:

**in utility.js : promptReactions, line 2935:**
![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/397e39ba36e656493f55514a021cd4ea2cda35e00487209dd9e7a691daa11093/image.png)

my bonusFlags.length is one but my triggerType is not "reaction" it is reactiondamage so it does not enter that code for optional.ac

**Comment by tposney** [comment]
Thanks for the work at digging into the error and you're right there is a typo.
The line in question should be 
```js
    reactionCount = reactionCount + Object.keys(midiFlags?.optional ?? [])

```

Also there are a couple of issues with the way the reaction is setup - here's the example I used which works after the above change is implemented. You can of course change NAME to be whatever you want.

![Screenshot_from_2023-01-03_12-37-54](https://my-gitlab-bucket.s3.amazonaws.com/640711674/03ae9b41fb89540887f9be3482bbdb16f462a69382f17cb67d85d7b923b4f66c/Screenshot_from_2023-01-03_12-37-54.png)

Fix in midi 10.0.24
